### ISSUE TITLE
Importing an image into C++\winRT WRC project

### ISSUE THREAD

**Issue by kennethgilpin** [microsoft/Win2D#714]
Sorry to ask what should be such a simple question.  I've tried for hours to no avail:

I am writing a win2D project in C++\winRT and I'm trying to import a .png file to generate a CanvasBitmap to display onto the screen during a canvas drawing session.  The image is called 'brain.png' and is embedded in a folder called 'Resources' within the project.

I think I'm creating the URI sensibly:

`// The URI string.`
`std::wstring uriImage{ L"ms-appx:///Resources/brain.png" };`
`// Create a Uri object from the URI string.`
`Windows::Foundation::Uri uri{ uriImage };`

But thereafter i'm struggling with the asynchronous request:
`CanvasBitmap brainBitmap = co_await CanvasBitmap::LoadAsync(*canvasDevice, uri);`

If I put the lot into IAsyncAction function and call with:
'auto async = callAsyncFunction();'

I get an 'Exception thrown: read access violation.' error.

I've read the posts on calling async requests from c++/winRT:
[https://kennykerr.ca/2018/03/01/cppwinrt-understanding-async/](url)
[https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/concurrency](url)

But I can't even load an image! :'(

Could someone point me in the right direction?  Async operations seem so opaque in C++/winRT compared to other languages.

With grateful thanks,  Kenneth

**Comment by kennethgilpin** [comment]
Just answering my own question here for the benefit of others:

Firstly **the URI of an embedded image ignores any folders it is in** - I'm guessing embedded files are put in the root of the application:
```
std::wstring uriImage{ L"ms-appx:///brain.png" };
Windows::Foundation::Uri uri{ uriImage };
```

Secondly even if you are loading an embedded asset, you still need to **enable the 'Pictures Library'**  checkbox in Package.appxmanifest > Capabilities

I think that both of these are probably bugs - they certainly aren't in the documentation

Here's a completed header file that loads SoftwareBitmaps and BitmapImages from URI and Files:

```
#pragma once

#include <winrt/Windows.Foundation.h>
#include <winrt/Windows.Storage.h>
#include <winrt/Windows.Storage.Streams.h>
#include <winrt/Windows.UI.Xaml.Media.Imaging.h>
#include <winrt/Windows.Graphics.Imaging.h>

namespace BitmapLoader
{
	using namespace winrt::Windows::Foundation;
	using namespace winrt::Windows::Storage;
	using namespace winrt::Windows::Storage::Streams;
	using namespace winrt::Windows::UI::Xaml::Media::Imaging;
	using namespace winrt::Windows::Graphics::Imaging;
	
	static IAsyncOperation<BitmapImage> GetBitmapImageFromURIasync(Uri uri)
	{
		RandomAccessStreamReference streamRef = RandomAccessStreamReference::CreateFromUri(uri);
		IRandomAccessStream stream{ co_await streamRef.OpenReadAsync() };
		BitmapImage bitmap{};
		bitmap.SetSource(stream);
		co_return bitmap;
	}
	static IAsyncOperation<BitmapImage> GetBitmapImageFromFileAsync(StorageFile file)
	{
		IRandomAccessStream stream{ co_await file.OpenAsync(FileAccessMode::Read) };
		BitmapImage bitmap{};
		bitmap.SetSource(stream);
		co_return bitmap;
	}

	static IAsyncOperation<SoftwareBitmap> GetSoftwareBitmapFromURIasync(Uri uri)
	{
		RandomAccessStreamReference streamRef = RandomAccessStreamReference::CreateFromUri(uri);
		IRandomAccessStream stream{ co_await streamRef.OpenReadAsync() };

		// Create the decoder from the stream
		BitmapDecoder decoder = co_await BitmapDecoder::CreateAsync(stream);

		// Get the SoftwareBitmap representation of the file
		SoftwareBitmap softwareBitmap = co_await decoder.GetSoftwareBitmapAsync();


		//convert to Bgra8 BitmapAlphaMode.Straight
		if (softwareBitmap.BitmapPixelFormat() != BitmapPixelFormat::Bgra8 ||
			softwareBitmap.BitmapAlphaMode() == BitmapAlphaMode::Straight)
		{
			softwareBitmap = SoftwareBitmap::Convert(softwareBitmap, BitmapPixelFormat::Bgra8, BitmapAlphaMode::Premultiplied);
		}

		co_return softwareBitmap;
	}
}
```

you might call these files from an AsyncMethod:
```
Windows::Foundation::IAsyncAction loadBitmapsAsync(
{
	std::wstring uriImage{ L"ms-appx:///brain.png" };
	Windows::Foundation::Uri uri{ uriImage };
	softwareBitmap = co_await BitmapLoader::GetSoftwareBitmapFromURIasync(uri);
}
```
which you might in turn call from the main thread:
```
auto processOp{ loadBitmapsAsync() };
```

**Comment by AndreasMiller** [comment]
Hi, thanks for sharing your question and answer. It is beautifully written code, by the way. I'm new to C++/winrt, and I have long been absent from the Microsoft/Windows environment. I had the same questions with respect to loading bitmaps, etc., and I got quite intimidated by all the ASync stuff. However, you have illustrated a very clear example of proper useage. Also, I tested your code and it works perfectly. Thanks!
